File: app/src/main/java/com/cookplan/recipe_import/approve_result/ImportRecipePresenterImpl.java
package com.cookplan.recipe_import.approve_result;

import com.cookplan.R;
import com.cookplan.RApplication;
import com.cookplan.models.CookPlanError;
import com.cookplan.models.Ingredient;
import com.cookplan.models.Product;
import com.cookplan.models.Recipe;
import com.cookplan.providers.IngredientProvider;
import com.cookplan.providers.ProductProvider;
import com.cookplan.providers.RecipeProvider;
import com.cookplan.providers.impl.IngredientProviderImpl;
import com.cookplan.providers.impl.ProductProviderImpl;
import com.cookplan.providers.impl.RecipeProviderImpl;
import com.cookplan.recipe_import.parser.Parser;
import com.cookplan.recipe_import.parser.ParserFactory;
import com.cookplan.recipe_import.parser.ParserResultListener;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import io.reactivex.SingleObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by DariaEfimova on 08.06.17.
 */

public class ImportRecipePresenterImpl implements ImportRecipePresenter {

    private ImportRecipeView mainView;
    private ProductProvider productDataProvider;
    private RecipeProvider recipeDataProvider;
    private IngredientProvider ingredientDataProvider;
    private CompositeDisposable disposables;
    private List<Product> allProductsList;

    public ImportRecipePresenterImpl(ImportRecipeView mainView) {
        this.mainView = mainView;
        productDataProvider = new ProductProviderImpl();
        recipeDataProvider = new RecipeProviderImpl();
        ingredientDataProvider = new IngredientProviderImpl();
        disposables = new CompositeDisposable();
    }

    @Override
    public void importRecipeFromUrl(String url) {
        getAsyncProductList();
        Parser parser = ParserFactory.createParser(url);
        if (parser != null) {
            parser.parceUrl(new ParserResultListener() {
                @Override
                public void onSuccess(Recipe recipe, Map<String, List<Ingredient>> ingredientList) {
                    if (mainView != null) {
                        mainView.setImportResult(recipe, ingredientList);
                    }
                }

                @Override
                public void onError(String error) {
                    if (mainView != null) {
                        mainView.setError(error);
                    }
                }
            });
        } else {
            if (mainView != null) {
                mainView.setError(
                        RApplication.getAppContext()
                                .getString(R.string.error_import_from_the_wrong_site_title));
            }
        }
    }

    @Override
    public List<Product> getAllProductsList() {
        return allProductsList;
    }

    private void getAsyncProductList() {
        disposables.add(
                productDataProvider.getProductList()
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableObserver<List<Product>>() {
                            @Override
                            public void onNext(List<Product> products) {
                                disposables.clear();
                                allProductsList = new ArrayList<>(products);
                            }

                            @Override
                            public void onError(Throwable e) {
                                allProductsList = new ArrayList<>();
                            }

                            @Override
                            public void onComplete() {

                            }
                        }));
    }

    @Override
    public void saveRecipe(Recipe recipe) {
        recipeDataProvider.createRecipe(recipe)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new SingleObserver<Recipe>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onSuccess(Recipe recipe) {
                        if (mainView != null) {
                            mainView.setRecipeSavedSuccessfully(recipe.getId());
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        if (mainView != null && e instanceof CookPlanError) {
                            mainView.setError(e.getMessage());
                        }
                    }
                });
    }
}

File: app/src/main/java/com/cookplan/recipe/import_recipe/approve_result/ImportRecipePresenterImpl.kt
package com.cookplan.recipe.import_recipe.approve_result

import com.cookplan.R
import com.cookplan.RApplication
import com.cookplan.models.*
import com.cookplan.providers.IngredientProvider
import com.cookplan.providers.ProductProvider
import com.cookplan.providers.ProviderFactory
import com.cookplan.providers.RecipeProvider
import com.cookplan.recipe.import_recipe.parser.ParserFactory
import com.cookplan.recipe.import_recipe.parser.ParserResultListener
import com.cookplan.utils.Utils
import com.google.firebase.auth.FirebaseAuth
import io.reactivex.SingleObserver
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers

/**
 * Created by DariaEfimova on 08.06.17.
 */

class ImportRecipePresenterImpl(private val mainView: ImportRecipeView?) : ImportRecipePresenter {
    private val productDataProvider: ProductProvider
    private val recipeDataProvider: RecipeProvider
    private val ingredientDataProvider: IngredientProvider
    private val disposables: CompositeDisposable
    private var allProductsList: List<Product> = listOf()
    private var recipeId: String? = null

    init {
        productDataProvider = ProviderFactory.Companion.productProvider
        recipeDataProvider = ProviderFactory.Companion.recipeProvider
        ingredientDataProvider = ProviderFactory.Companion.ingredientProvider
        disposables = CompositeDisposable()
    }

    override fun importRecipeFromUrl(uri: String) {
        getAsyncProductList()
        ParserFactory.createParser(uri)?.parceUrl(object : ParserResultListener {
            override fun onSuccess(recipe: Recipe, ingredientList: MutableMap<String, List<Ingredient>>) {
                mainView?.setImportResult(recipe, ingredientList)
            }

            override fun onError(error: String) {
                mainView?.setError(error)
            }
        }) ?: mainView?.setError(
                RApplication.appContext!!
                        .getString(R.string.error_import_from_the_wrong_site_title))
    }

    override fun getAllProductsList() = allProductsList

    private fun getAsyncProductList() {
        disposables.add(
                productDataProvider.getProductList()
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableObserver<List<Product>>() {
                            override fun onNext(products: List<Product>) {
                                disposables.clear()
                                allProductsList = products
                            }

                            override fun onError(e: Throwable) {
                            }

                            override fun onComplete() {

                            }
                        }))
    }

    override fun saveRecipe(recipe: Recipe) {
        recipeDataProvider.createRecipe(recipe)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : SingleObserver<Recipe> {
                    override fun onSubscribe(d: Disposable) {

                    }

                    override fun onSuccess(recipe: Recipe) {
                        recipeDataProvider.getRecipeById(recipe.id)
                                .subscribeOn(Schedulers.io())
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribe(object : SingleObserver<Recipe> {
                                    override fun onSubscribe(d: Disposable) {

                                    }

                                    override fun onSuccess(recipe: Recipe) {
                                        recipeId = recipe.id
                                        mainView?.setRecipeSavedSuccessfully(recipe)
                                    }

                                    override fun onError(e: Throwable) {
                                        mainView?.setError(e.message ?: String())
                                    }
                                })
                    }

                    override fun onError(e: Throwable) {
                        mainView?.setError(e.message ?: String())
                    }
                })
    }

    override fun saveProductAndIngredient(key: String, category: ProductCategory, name: String, amount: Double, measureUnit: MeasureUnit) {
        var map: Map<MeasureUnit, Double> = mapOf()
        var measureUnitList: MutableList<MeasureUnit> = MeasureUnit.values().toMutableList()
        if (measureUnit === MeasureUnit.KILOGRAMM) {
            map = Utils.kilogramUnitMap
            measureUnitList = Utils.weightUnitList
        }
        if (measureUnit === MeasureUnit.LITRE) {
            map = Utils.litreUnitMap
            measureUnitList = Utils.volumeUnitList
        }
        var rusName: String? = null
        var engName: String? = null
        if (RApplication.isCurrentLocaleRus) {
            rusName = name
        } else {
            engName = name
        }
        val product = Product(category, rusName, engName, mutableListOf(measureUnit),
                measureUnitList, map, FirebaseAuth.getInstance().currentUser?.uid)
        product.increasingCount()
        productDataProvider.createProduct(product)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : SingleObserver<Product> {
                    override fun onSubscribe(d: Disposable) {

                    }

                    override fun onSuccess(product: Product) {
                        //save ingredient
                        val ingredient = Ingredient(null, product.toStringName(), product, recipeId,
                                measureUnit, amount, ShopListStatus.NONE)
                        saveIngredient(key, ingredient)
                    }

                    override fun onError(e: Throwable) {
                        mainView?.setError(e.message ?: String())
                    }
                })
    }

    override fun saveIngredient(key: String, ingredient: Ingredient) {
        ingredientDataProvider.createIngredient(ingredient)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : SingleObserver<Ingredient> {
                    override fun onSubscribe(d: Disposable) {

                    }

                    override fun onSuccess(ingredient: Ingredient) {
                        mainView?.setIngredientSavedSuccessfully(key)
                    }

                    override fun onError(e: Throwable) {
                        mainView?.setError(e.message ?: String())
                    }
                })
    }
}